### PR TITLE
feat(nuget): add support for updating lock files

### LINKED ISSUE #6610: Update NuGet Lock file
**What would you like Renovate to be able to do?**
Renovate should update the NuGet lock file when updating NuGet references

**Describe the solution you'd like**
NuGet supports using a "lock file" that enables downloading the *exact* same package for every build/restore.
For CI builds, the NuGet documentation recommends to use the "locked mode" which causes the build to fail if the lock file is not up to date with a project's package references.

So to avoid failing CI builds when a lock file is used, Renovate should also update the lock file in addition to a project's package references.

**Describe alternatives you've considered**
There are different ways to work around it:

- Disable usage of the NuGet lock file (it is optional). However, this might lead to different builds downloading different versions of a package. This is especially important when using private NuGet feeds. Because NuGet.org does not allow deletion or modification of packages, I don't think it can happen when packages are downloaded from there.
- It is also possible to work around this by having a CI build that updates the lock file on branches created by Renovate (for example, I've set up something like this using GitHub Actions). However, I think it would be better if this was integrated in Renovate because Renovate will stop updating a branch once it has additional commits.

**Additional context**

- [Enable repeatable package restores using a lock file (NuGet Blog)](https://devblogs.microsoft.com/nuget/enable-repeatable-package-restores-using-a-lock-file/)

### PR DESCRIPTION
**Pull request by fgreinacher**
Resolves #6610

### CODE REVIEW

**Comment by fgreinacher** [comment]
OK, this is ready for review now @viceice

**Comment by fgreinacher** [comment]
(And I think we need to wait for the Docker image, right?)

**Comment by viceice** [comment]
Yes, sorry.

**Comment by fgreinacher** [comment]
> Yes, sorry.

Not meant as offense, just wasn't sure whether something happened that I have not noticed 🤣

**Comment by viceice** [comment]
Yes, too much work on daily job happened. 🙈

File: lib/manager/nuget/artifacts.ts [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+export async function updateArtifacts({
+  packageFileName,
+  newPackageFileContent,
+  config,
+  updatedDeps,
+}: UpdateArtifact): Promise<UpdateArtifactsResult[] | null> {
+  logger.debug(`nuget.updateArtifacts(${packageFileName})`);
+
+  if (!/[cs|vb|fs]proj$/i.test(packageFileName)) {
```
Comment (JamieMagee) [inline lib/manager/nuget/artifacts.ts:67]:
```suggestion
  if (!/(?:cs|vb|fs)proj$/i.test(packageFileName)) {
```

Same as the filematch regex

https://github.com/renovatebot/renovate/blob/f52fe033de5fae6be0155c3b36fad1ec6f722f01/lib/manager/nuget/index.ts#L8-L12

File: lib/manager/nuget/artifacts.ts [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    // This could be implemented in the future if necessary.
+    // It's not that easy though because the questions which
+    // project file to restore how to detmermine which lock files
```
Comment (JamieMagee) [inline lib/manager/nuget/artifacts.ts:70]:
```suggestion
    // project file to restore how to determine which lock files
```

**Review by viceice — COMMENTED** [review]
LGTM, but needs dotnet docker images to be published first

**Comment by viceice** [comment]
https://hub.docker.com/r/renovate/dotnet/tags

🎉

**Comment by rarkins** [comment]
Can someone verify that it works with Docker too?

**Review by rarkins — CHANGES_REQUESTED** [review]
Looks ok from a code review point of view, should be tested with Docker first before merging

**Comment by fgreinacher** [comment]
> Looks ok from a code review point of view, should be tested with Docker first before merging

I'll take care!

**Comment by fgreinacher** [comment]
> > Looks ok from a code review point of view, should be tested with Docker first before merging
> 
> I'll take care!

Found a bug (fixed with https://github.com/renovatebot/renovate/pull/7375/commits/68239c6ea99f0433298dda4476e6258567c87c11). 

Otherwise it's working with both [Global](https://github.com/fgreinacher/nuget-lockfile/pull/1/files) and [Docker](https://github.com/fgreinacher/nuget-lockfile-2/pull/1/files) binary sources.

**Comment by fgreinacher** [comment]
(BTW: I first tried this on Windows, where I ran into several problems with the Docker binary source. I ignored this and switched to my Mac, assuming that it's OK like that. Tell me if this should actually work, I'd then dig a bitter deeper to see if I can write a proper issue)

**Comment by viceice** [comment]
> (BTW: I first tried this on Windows, where I ran into several problems with the Docker binary source. I ignored this and switched to my Mac, assuming that it's OK like that. Tell me if this should actually work, I'd then dig a bitter deeper to see if I can write a proper issue)

This will never work on windows because the host and container path need to match. A workaround is to mounbt the local build into the renovate slim image and use that to run renovate in container. Then you are able to mount /tmp to renovate, which will be used as shared volume

**Comment by fgreinacher** [comment]
> > (BTW: I first tried this on Windows, where I ran into several problems with the Docker binary source. I ignored this and switched to my Mac, assuming that it's OK like that. Tell me if this should actually work, I'd then dig a bitter deeper to see if I can write a proper issue)
> 
> This will never work on windows because the host and container path need to match. A workaround is to mounbt the local build into the renovate slim image and use that to run renovate in container. Then you are able to mount /tmp to renovate, which will be used as shared volume

Yeah, that's why I just switched to a proper machine 😁. Just wanted to make sure I didn't miss anything essential.

**Comment by viceice** [comment]
Alternate would be to use wsl on windows 😉

**Review by viceice — APPROVED** [review]

**Pull request merged**